src/components/input/input.ts

```diff
@@ -0,0 +1,113 @@
+import { customElement, LitElement, html, property } from 'lit-element';
+import classnames from 'classnames';
+import settings from 'carbon-components/es/globals/js/settings';
+import WarningFilled16 from '@carbon/icons/lib/warning--filled/16';
+import styles from './input.scss';
+
+const { prefix } = settings;
+
+/**
+ * Supported input types.
+ *
+ * For this component we only support textual types
+ */
+export enum INPUT_TYPE {
+  EMAIL = 'email',
+  PASSWORD = 'password',
+  TEL = 'tel',
+  TEXT = 'text',
+  URL = 'url',
+}
+
+/**
+ * Input element. Supports all the usual attributes for textual input types
+ */
+@customElement(`${prefix}-input`)
+export default class BXInput extends LitElement {
+  @property()
+  autocomplete = '';
+
+  @property({ type: Boolean })
+  autofocus = false;
+
+  @property({ type: Boolean, reflect: true })
+  disabled = false;
+
+  @property()
+  form = '';
+
+  @property({ type: Boolean, reflect: true })
+  invalid = false;
+
+  @property()
+  name = '';
+
+  @property()
+  pattern = '';
+
+  @property({ reflect: true })
+  placeholder = '';
+
+  @property({ type: Boolean, reflect: true })
+  readonly = false;
+
+  @property({ type: Boolean, reflect: true })
+  required = false;
+
+  @property({ reflect: true })
+  type = INPUT_TYPE.TEXT;
+
+  @property({ reflect: true })
+  value = '';
+
+  /**
+   * Unique ID used for ID refs.
+   */
+  protected _id = Math.random()
+    .toString(36)
+    .slice(2);
+
+  render() {
+    const invalidIcon = WarningFilled16({ class: 'bx--text-input__invalid-icon' });
+
+    const inputClasses = classnames(`${prefix}--text-input`, {
+      [`${prefix}--text-input--invalid`]: this.invalid,
+    });
+
+    const labelClasses = classnames(`${prefix}--label`, {
+      [`${prefix}--label--disabled`]: this.disabled,
+    });
+
+    const helperTextClasses = classnames(`${prefix}--form__helper-text`, {
+      [`${prefix}--form__helper-text--disabled`]: this.disabled,
+    });
+
+    return html`
+      <label class="${labelClasses}" for="${this._id}"><slot name="label"></slot></label>
+      <div class="${helperTextClasses}"><slot name="help-text"></slot></div>
+      <div class="${prefix}--text-input__field-wrapper" ?data-invalid="${this.invalid}">
+        ${this.invalid ? invalidIcon : null}
+        <input
+          ?autocomplete="${this.autocomplete}"
+          ?autofocus="${this.autofocus}"
+          class="${inputClasses}"
+          ?data-invalid="${this.invalid}"
+          ?disabled="${this.disabled}"
+          form="${this.form}"
+          id="${this._id}"
```

Comment:
Wondering if the ID there can be based on the host element ID, like https://github.com/carbon-design-system/carbon-custom-elements/blob/bc73ab5/src/components/dropdown/dropdown.ts#L88-L94. Alternatively, we can do something like just `input` given each shadow tree has unique ID namespace.